### ISSUE TITLE
No error when redeclaring a symbol twice with two incompatible types

### ISSUE THREAD

**Issue by lukaslihotzki-f** [microsoft/pyright#10173]
As per https://github.com/microsoft/pyright/issues/10170#issuecomment-2754744412, declaring classes with the same name in if/else blocks is redeclaring a symbol twice with two incompatible types. Therefore, pyright needs to detect a type error.

**Code or Screenshots**
The following code should generate a type error (`error: Class declaration "Foo" is obscured by a declaration of the same name (reportRedeclaration)`), but no errors or warnings are output:

```python
if True:
    class Foo:
        pass
else:
    class Foo:
        pass
```

**VS Code extension or command-line**
I reproduced this issue in the command line tool, using pyright 1.1.397.

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug.

Type checkers like pyright statically evaluate a small subset of conditional expressions at binding time. This is mandated by the Python typing spec to handle checks like `if sys.version >= (3, 9)`, and many libraries and stubs rely on this behavior. Pyright also handles `True` and `False` since they're built-in keywords. Based on this, it treats the `else` clause in your code above as never reachable, so this isn't considered a redeclaration.

**Comment by lukaslihotzki-f** [comment]
@erictraut, thanks for the detailed explanations.
